sentry/sentry.go

```diff
@@ -1,71 +1,481 @@
 package sentry
 
 import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
 	"net/http"
 	"net/url"
-	"path"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
 
-	"github.com/dghubble/sling"
+	"github.com/google/go-querystring/query"
+	"github.com/peterhellberg/link"
 )
 
-// public constants for external use
 const (
-	DefaultBaseURL = "https://sentry.io/api/"
-	APIVersion     = "0"
+	defaultBaseURL = "https://sentry.io/api/"
+	userAgent      = "go-sentry"
+
+	// https://docs.sentry.io/api/ratelimits/
+	headerRateLimit               = "X-Sentry-Rate-Limit-Limit"
+	headerRateRemaining           = "X-Sentry-Rate-Limit-Remaining"
+	headerRateReset               = "X-Sentry-Rate-Limit-Reset"
+	headerRateConcurrentLimit     = "X-Sentry-Rate-Limit-ConcurrentLimit"
+	headerRateConcurrentRemaining = "X-Sentry-Rate-Limit-ConcurrentRemaining"
 )
 
-// Client for sentry api
+var errNonNilContext = errors.New("context must be non-nil")
+
+// Client for Sentry API.
 type Client struct {
-	sling               *sling.Sling
-	Organizations       *OrganizationService
-	OrganizationMembers *OrganizationMemberService
-	Teams               *TeamService
-	Projects            *ProjectService
-	ProjectKeys         *ProjectKeyService
-	ProjectPlugins      *ProjectPluginService
-	Rules               *RuleService
-	Ownership           *ProjectOwnershipService
+	client *http.Client
+
+	// BaseURL for API requests.
+	BaseURL *url.URL
+
+	// User agent used when communicating with Sentry.
+	UserAgent string
+
+	// Latest rate limit
+	rate Rate
+
+	// Common struct used by all services.
+	common service
+
+	// Services
+	DashboardWidgets    *DashboardWidgetsService
+	Dashboards          *DashboardsService
+	IssueAlerts         *IssueAlertsService
+	MetricAlerts        *MetricAlertsService
+	OrganizationMembers *OrganizationMembersService
+	Organizations       *OrganizationsService
+	ProjectKeys         *ProjectKeysService
+	ProjectOwnerships   *ProjectOwnershipsService
+	ProjectPlugins      *ProjectPluginsService
+	Projects            *ProjectsService
+	Teams               *TeamsService
+	ProjectFilter       *ProjectFilterService
```

Comment:
Add a new line before our struct members; in this way, the padding will be independent of upstream, reduce changes, and make it easier to manage conflicts.

Reply:
Nice catch! Added in 0631526d072a76ad479eca76c1eee4f2519cef87